pkg/cmd/cli/cmd/login.go

```diff
@@ -0,0 +1,248 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+
+	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	kclientcmd "github.com/GoogleCloudPlatform/kubernetes/pkg/client/clientcmd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/cli/config"
+	"github.com/openshift/origin/pkg/cmd/util"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+)
+
+type loginOptions struct {
+	username string
+	password string
+	server   string
+	project  string
+	context  string
+}
+
+func NewCmdLogin(f *osclientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &loginOptions{}
+
+	cmds := &cobra.Command{
+		Use:   "login [--username=<username>] [--password=<password>] [--server=<server>]",
+		Short: "Logs in and returns a session token",
+		Long: `Logs in to the OpenShift server and prints out a session token.
+
+Username and password can be provided through flags, the command will
+prompt for user input if not provided.
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			// fetch config and if a server were not provided, prompt for it
+			clientCfg, err := f.OpenShiftClientConfig.ClientConfig()
+			if isServerNotFound(err) {
+				defaultServer := osclientcmd.DefaultClusterURL
```

Comment:
> > +			if isServerNotFound(err) {
> > +				defaultServer := osclientcmd.DefaultClusterURL
> 
> This value should be on the clientCfg as Host... You don't need to make
> defaultClusterURL public.

It's not in clientCfg as Host at this point because we don't have a default host anymore (our default host is ""). We don't default to a host because we require users to always provide it. I moved DefaultClusterURL from here to the `login` command where it's now only used to display "or just hit <enter> to use https://localhost:8443:".

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/992/files#r25139115